### ISSUE TITLE
Build instructions for Debian missing / incomplete

### ISSUE THREAD

**Issue by mYnDstrEAm** [xbmc/xbmc#18612]
### Describe the bug
There is a build guide for Ubuntu but not for Debian and the former does not work with Debian. **[The Linux build guide](https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md)** is incomplete with Debian.

## Expected Behavior
Clear and easy steps in the guide. Users should not be required to build Kodi on Debian (there is a version of Kodi in Debian's repos) but _if_ they have to the steps should be complete and simple.

## Actual Behavior
The Linux guide was incomplete.

## Possible Fix
I had to run `sudo apt-get install rapidjson-dev`

> /usr/bin/ld: cannot find -lunistring

can be fixed by running this before the build command:
`sudo apt-get install libunistring-dev`

The list of required packages is in a format one can't really use: there should be a command you can just copy & paste in the guide like (without [armel]):

`sudo apt-get install autoconf automake autopoint gettext autotools-dev cmake curl default-jre openjdk-6-jre openjdk-7-jre gawk gcc gcc-4.9 g++ g++-4.9 cpp cpp-4.9 flatbuffers gdc gperf libasound2-dev libasound-dev libass-dev libavahi-client-dev libavahi-common-dev libbluetooth-dev libbluray-dev libbz2-dev libcdio-dev libcec4-dev libcec-dev libp8-platform-dev libcrossguid-dev libcurl4-openssl-dev libcurl4-gnutls-dev libcurl-dev libcwiid-dev libdbus-1-dev libegl1-mesa-dev libenca-dev libflac-dev libfontconfig-dev libfmt3-dev libfmt-dev libfreetype6-dev libfribidi-dev libfstrcmp-dev libgcrypt-dev libgif-dev libgles2-mesa-dev libgl1-mesa-dev libgl-dev libglew-dev libglu1-mesa-dev libglu-dev libgnutls-dev libgnutls28-dev libgpg-error-dev libgtest-dev libiso9660-dev libjpeg-dev liblcms2-dev liblirc-dev libltdl-dev liblzo2-dev libmicrohttpd-dev libmysqlclient-dev libnfs-dev libogg-dev libomxil-bellagio-dev libpcre3-dev libplist-dev libpng12-dev libpng-dev libpulse-dev libshairplay-dev libsmbclient-dev libspdlog-dev libsqlite3-dev libssl-dev libtag1-dev libtag1x8 libtiff5-dev libtiff-dev libtiff4-dev libtinyxml-dev libtool libudev-dev libva-dev libvdpau-dev libvorbis-dev libxkbcommon-dev libxmu-dev libxrandr-dev libxslt1-dev libxslt-dev libxt-dev waylandpp-dev netcat wayland-protocols wipe lsb-release meson nasm ninja-build python3-dev python3-pil python-imaging python-support python3-minimal rapidjson-dev swig unzip uuid-dev yasm zip zlib1g-dev`

Furthermore it would be good if "Enable internal dependencies" was clarified a bit better and/or marked as optional.

Is usage of the Krypton branch intentional? If not this should be updated and maybe also have the option to use the releases page.

When running `make check` I get:

`
[ 95%] Building CXX object CMakeFiles/kodi-test.dir/xbmc/video/test/TestVideoInfoScanner.cpp.o
/home/username/xbmc-19.0a2-Matrix/xbmc/video/test/TestVideoInfoScanner.cpp:64:25: error: expected constructor, destructor, or type conversion before ‘(’ token
 INSTANTIATE_TEST_SUITE_P(VideoInfoScanner, TestVideoInfoScanner, ValuesIn(TestData));
                         ^
make[3]: *** [CMakeFiles/kodi-test.dir/build.make:89: CMakeFiles/kodi-test.dir/xbmc/video/test/TestVideoInfoScanner.cpp.o] Fehler 1
make[2]: *** [CMakeFiles/Makefile2:1058: CMakeFiles/kodi-test.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:405: CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:305: check] Error 2
`

Lastly (and this might be a separate issues but is easy to fix): can you please provide hash sums for the releases? I don't know why this is not getting done for all releases (not just Kodi) as a standard routine.Please jJust run e.g. `sha256sum 19.0a2-Matrix.zip`and add the hash to the guide or the releases page.

### To Reproduce
1. Install Debian 10 stable if not installed already
2. Try to install the latest Kodi without using the flatpak

### Debuglog
Not relevant here

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Debian 10 stable
 - Kodi version: 19 alpha 2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by basilgello** [comment]
Hi @mYnDstrEAm !

In Debian, build-dependencies can be easily installed using `apt build-dep kodi` and the Debian source package can be fetched from Debian mirrors using `apt source kodi`. This applies, however, to a release present in Debian (currently 18.8) and requires uncommenting `deb-src` line in `/etc/apt/sources.list` file. Debian releases are signed by Debian developer's GPG key and validated by apt / dpkg automagically after source tarball is diwnloaded and before is unpacked.

However, your issue is totally valid in a usecase when an end-user is willing to build newest Kodi that is not yet present in Debian. I am the new Kodi maintainer in Debian and I will update the build instructions after I push the Kodi mirror to experimental. In the meantime, you can use my unofficial repository: https://basilgello.github.io/kodi-nightly-debian-repo

**Comment by mYnDstrEAm** [comment]
> In Debian, build-dependencies can be easily installed using `apt build-dep kodi` and the Debian source package can be fetched from Debian mirrors using `apt source kodi`. This applies, however, to a release present in Debian (currently 18.8) and requires uncommenting `deb-src` line in `/etc/apt/sources.list` file.

Then this should probably be part of the guide. However 18.8 is only part of sid and mixing different releases [is discouraged](https://wiki.debian.org/DontBreakDebian#Don.27t_make_a_FrankenDebian). Maybe this is an exception - what would be the recommended way here and why? Relevant StackExchange question [here](https://unix.stackexchange.com/questions/584798/how-to-install-latest-kodi-on-latest-debian).

>Debian releases are signed by Debian developer's GPG key and validated by apt / dpkg automagically after source tarball is diwnloaded and before is unpacked.

This is fine but the distribution from the Kodi devs to the Debian devs as well as from the Kodi devs to the public seems to still occur without hashsums even though fixing this would require only 2 minutes and has clear advantages.

> However, your issue is totally valid in a usecase when an end-user is willing to build newest Kodi that is not yet present in Debian. I am the new Kodi maintainer in Debian and I will update the build instructions after I push the Kodi mirror to experimental. In the meantime, you can use my unofficial repository: https://basilgello.github.io/kodi-nightly-debian-repo

Thank you for commenting here! That repo is very useful - I was looking for something like it in the StackExchange question linked above (you could also post an answer there or I'll edit mine).  
I'm hoping that backports gets changed and/or some new mechanism gets build that allows for installing newer version of specific software as described [here](https://unix.stackexchange.com/questions/584892/what-would-be-needed-by-debian-to-ship-newer-versions-of-the-kde-desktop-environ). If you know a way such could be implemented or any other relevant info, please let me know. One way I could think of would be making repositories with newer versions like yours and Preining's repo for a newer version of KDE "official" and maybe to integrate them with the main repo so that, for example, you can choose which version of a package to install (the default would remain as is and GUIs like Apper would display all available options including the version in backports).  
(Outdated software is a main concern/drawback about Debian - I know it's also part of its way to ensure stability, which is one of it's core strengths but resolving it in a good way would allow more people to install Debian as well as enabling better security and convenience for those who already have Debian stable installed and would like to use newer versions of specific software.)

**Comment by basilgello** [comment]
> you could also post an answer there or I'll edit mine

Please edit yours as I have no SO account :)

> Then this should probably be part of the guide. However 18.8 is only part of sid and mixing different releases is discouraged. Maybe this is an exception - what would be the recommended way here and why? Relevant StackExchange question here.

I will rewrite that guide in Debian part.

> I'm hoping that backports gets changed and/or some new mechanism gets build that allows for installing newer version of specific software as described here. If you know a way such could be implemented or any other relevant info, please let me know. One way I could think of would be making repositories with newer versions like yours and Preining's repo for a newer version of KDE "official" and maybe to integrate them with the main repo so that, for example, you can choose which version of a package to install (the default would remain as is and GUIs like Apper would display all available options including the version in backports).
(Outdated software is a main concern/drawback about Debian - I know it's also part of its way to ensure stability, which is one of it's core strengths but resolving it in a good way would allow more people to install Debian as well as enabling better security and convenience for those who already have Debian stable installed and would like to use newer versions of specific software.)

Once Kodi 19.0 lands in unstable and then in testing, I will push buster-backports. This way you will need just enable it in your sources list and install with `apt install -t buster-backports kodi`. It is totally encouraged practice, yet not enabled by default.

Sources are hosted on my personal Debian Salsa space: https://salsa.debian.org/basilgello. I will add all binary addons there this week, so the repository in Debian will be feature-par with Team Kodi's PPA.

**Comment by wsnipex** [comment]
> This is fine but the distribution from the Kodi devs to the Debian devs as well as from the Kodi devs to the public seems to still occur without hashsums even though fixing this would require only 2 minutes and has clear advantages.

What do you mean? There is no distribution from Kodi to debian.

And all our files from our mirror server do have sha256 hashsums, if you install e.g. an addon from within kodi, those hashsums are verified. Coupled with HTTPS this is good enough, even though gpg signatures would be better.

**Comment by basilgello** [comment]
> There is no distribution from Kodi to debian.

Right. Debian maintainers pick upstream tarballs from Github, repack them removing files not compliant to Debian Free Software Guidelines (DFSG), apply the necessary patches to make Kodi build and upload to Debian master FTP server. Master FTP server is queried by Debian builder machines (buildd-s aka https://buildd.debian.org) and the compiled DEBs are uploaded to Debian mirrors.

**Comment by mYnDstrEAm** [comment]
> What do you mean? There is no distribution from Kodi to debian.

Somehow the Debian's Kodi maintainers retrieve the Kodi software - that's what I was referring to.

> And all our files from our mirror server do have sha256 hashsums, if you install e.g. an addon from within kodi, those hashsums are verified. Coupled with HTTPS this is good enough, even though gpg signatures would be better.

Could you please just add the hashsums here? I did a sha256sum but could not find it. Solving it would be as simple as adding `aa537c5f0fcbb073021a919844623aaa1e33b102f2844a0c41d625ef3baf54cf` to the releases page. GPG signatures would be better. Ensuring integrity of Kodi addons, especially from unofficial repos, would be a separate issue and probably not as simple to resolve.

**Comment by wsnipex** [comment]
you can get the hashsum for all files on our mirrors like this:
https://mirrors.kodi.tv/addons/matrix/addons.xml.gz.sha256

**Comment by mYnDstrEAm** [comment]
Thanks, that's helpful! Maybe this site should be linked from the releases page? I noticed that in releases/source/ the Matrix releases are missing.  
Ensuring authenticity of distributed packages as well as their security is a large and separate issue.  
I was only asking about a short-term and very easy to implement improvement: adding the SHA256 sums to the releases page, like many other GitHub projects are already doing. Maybe I should create a separate issue for it? I didn't do so because it's so easy to solve and takes about an extra minute after the latest release is added.

[I'm using basilgello's repo for Kodi v19 now and it works nicely on Debian/KDE](https://github.com/basilgello/kodi-nightly-debian-repo/issues/1).

Could you please change the The Linux build guide by at least adding commands `sudo apt-get install rapidjson-dev` and `sudo apt-get install libunistring-dev` as explained above and if possible also another (copy-pastable) format of the list of required packages, a note about the "Enable internal dependencies" and an updated branch?

**Comment by wsnipex** [comment]
which releases page are you talking about exactly?

As for the build guide: please feel free to open a PR

**Comment by mYnDstrEAm** [comment]
[This page.](https://github.com/xbmc/xbmc/releases) Okay.

**Comment by wsnipex** [comment]
Not really, because there is guarantee that the checksum will not change on github. The reason is that git archive (which github uses to generate tarballs on the fly) is not guaranteed to always produce the same hash.
If you do your research you will find that is has bitten many other projects where build scripts relied on that checksum being always the same.

If you need that reliability, you can use the source from http://mirrors.kodi.tv/releases/source/

**Comment by mYnDstrEAm** [comment]
Hashsums are there to help ensure authenticity of packages and precisely that software does _not_ change via any means. (The exactly same software should land on users' storage for the same version and hashsums can be used to check for this. So this is the entire prupose of them.) The hashsum could be added to description after the tarball has been generated. Other projects also have this in their releases' descriptions.

**Comment by wsnipex** [comment]
you don't listen. Github/git archive doesn't guarantee to always produce the exact same tarball, nothing we can do about it.
Use the souce and checksum from our mirrors

**Comment by mYnDstrEAm** [comment]
Maybe we're misunderstanding each other: hashsums are there to ensure authenticity of packages. If they can't be used authenticity can not be ensured. But authenticity should obviously be ensured for security-reasons. If a site can't be used to guarantee it, it probably shouldn't be used. But as said, it seems like it can as other releases pages also have the hashsums of all files in the description. And shouldn't the tarball be produced only when a new release is added to the releases page? At this time (when the new release is added) simply also add a hashsum. Why not? (Also the mirrors don't even have the source of the v 19.0 release but that's not what this is about.)

**Comment by Razzeee** [comment]
I'm not getting, why you want it added to https://github.com/xbmc/xbmc/releases as @wsnipex is correct and the hashes can change (ask github why).

I would have understood wanting them on https://kodi.tv/download but then again, I'm not sure if we want to change the website right now.

**Comment by strugee** [comment]
Filed #18990 before I saw this issue. It fixes the `libunistring-dev` issue mentioned here.

As for hashes, there's no point in adding them to GitHub. My guess is that what @mYnDstrEAm is seeing in other projects is that they manually upload source code tarballs, in addition to the automatic "Source code" release artifacts that GitHub generates (which according to this thread can change in contents). Either that or other projects aren't aware that these files can change. I certainly wasn't.

That being said, however, adding hashes would _not_ "ensure authenticity of packages" - this is a common misconception. If you don't trust GitHub to not tamper with source code tarballs on the xbmc/xbmc Releases page, then you also don't trust GitHub to not tamper with the hashes on that page. Same goes for https://kodi.tv/download. Hashes ensure integrity but they do not ensure authenticity - to do that you would need some sort of signing scheme like Signify or (ugh) GPG.

**Comment by strugee** [comment]
Did some digging. The `INSTANTIATE_TEST_SUITE_P` error is caused by the version of gtest in Debian being too old. You can work around it by using the Kodi vendored gtest; just pass `-DENABLE_INTERNAL_GTEST=ON` when configuring CMake.

I will likely prepare a patch for the documentation for this tomorrow, but if someone else wants to do this (or more likely if I forget) please feel free to beat me to it!

**Comment by fritsch** [comment]
Can't you just update debian's gtest version? Introducing workarounds for religious philosophy is wrong by design. Or is newest gtest version unstable?

**Comment by basilgello** [comment]
C'on folks, I rebuilt everything for buster in my repo including gtest, flatbuffers etc. If you need instructions, I can explain how to do it.
-- 
Vasyl Gello
==================================================
Certified SolidWorks Expert

Mob.:+380 (98) 465 66 77

E-Mail: vasek.gello@gmail.com

Skype: vasek.gello
==================================================
호랑이는 죽어서 가죽을 남기고 사람은 죽어서 이름을 남긴다

**Comment by philiprgbel** [comment]
Get Outlook for iOS<https://aka.ms/o0ukef>




________________________________

**Comment by philiprgbel** [comment]
Thank you

**Comment by philiprgbel** [comment]
Anything else 
H
